### ISSUE TITLE
Python 3 Guest Worker stdout print

### ISSUE THREAD

**Issue by singlecheeze** [crossbario/crossbar#314]
crossbar/crossbar/controller/processtypes.py

Python 3 Guest worker will not print to router log from simple print statement. Could this have to do with twisted.python.log and Defered?:

```python
    self._log_topic = 'crossbar.node.{}.worker.{}.on_log'.format(self._controller._node_id, self.id)
    # A deferred that resolves when the worker is ready.
    self.ready = Deferred()

    # A deferred that resolves when the worker has exited.
    self.exit = Deferred()
```

Going for this functionality:
http://crossbar.io/docs/Guest-Configuration/
stdout: Action on signal on standard out, can be close, log or drop
stderr: Action on signal on standard error, can be close, log or drop

More references:
https://groups.google.com/forum/#!topic/crossbario/gt_JLPcFxmU

https://groups.google.com/forum/#!topic/autobahnws/A3yjexYaiPw

https://groups.google.com/forum/#!topic/autobahnws/3FRF7RRw8wo

**Comment by meejah** [comment]
I can confirm there's some weirdness with the guest logging: Python 3.4.2 guests printing nothing. Under Python 2.7.9 or PyPy (with or without trollius) the ``stderr`` messages are logged immediately, but stdout only after shutdown.

Passing ``stdin: close`` doesn't change the behavior. So, definitely a bug.

**Comment by meejah** [comment]
A simple bash script as a guest seems to log everything properly.

**Comment by meejah** [comment]
https://bugs.python.org/issue11633 and https://docs.python.org/2/using/cmdline.html#cmdoption-u are related.

If you set ``PYTHONUNBUFFERED=1`` in your config, logging works as-expected. Perhaps crossbar should just set this env-var automagically? To immediately fix the issue described, add a config snippit like this to your guest:

```json
"options": {
    "env": {
        "vars": {
            "PYTHONUNBUFFERED": 1
        }
    }
}
```

**Comment by singlecheeze** [comment]
Confirmed this workaround. Thank you @meejah so much!!!

**Comment by singlecheeze** [comment]
Crossbar config.json required a string of "1", it would not accept JSON Int 1

{
         "id": "backend",
         "type": "guest",
         "executable": "python3",
         "arguments": ["backend.py"],
         "options": {
            "env": {
               "vars": {
                  "PYTHONUNBUFFERED": "1"
               }
            },
            "workdir": "../backend",
            "stdout": "log",
            "stderr": "log",
            "watch": {
               "directories": ["../backend"],
               "action": "restart"
            }
         }
      }

**Comment by meejah** [comment]
You can also add ``"-u"`` to the ``arguments`` list (instead of the env-var).

**Comment by meejah** [comment]
@oberstet do you think crossbar's behavior should be changed, or is my PR #59 on crossbarwww sufficient? We could, e.g., add PYTHONBUFFERED for all guests, or only if some metric indicates it's a python process? This could be a little fragile, but might give a better "out of the box" user experience...

**Comment by oberstet** [comment]
@meejah for me, automatically setting env vars is too much magic. The docs should prominently mention it (they now do in a note at least thx to your patch), and our examples / templates should do it.
